### PR TITLE
KAFAK-3522: Add TopologyTestDriver unit tests

### PR DESCRIPTION
**Pull request by mjsax**
Part of KIP-258. 

- For backward compatibility, we need to return the correct store via IQ API
- Add additional unit test for #6175

### CODE REVIEW

**Comment by bbejeck** [comment]
@mjsax any updates on this PR as #6175 has been merged.

**Comment by mjsax** [comment]
Need to find some time to rebase this PR... Will ping you when it's ready. Thanks for following up.

File: streams/src/main/java/org/apache/kafka/streams/state/Stores.java [hunk 1]
```diff
@@ -353,26 +353,6 @@ public static WindowBytesStoreSupplier inMemoryWindowStore(final String name,
...
-    /**
-     * Create an in-memory {@link SessionBytesStoreSupplier}.
-     * @param name              name of the store (cannot be {@code null})
-     * @param retentionPeriod   length ot time to retain data in the store (cannot be negative)
-     *                          Note that the retention period must be at least long enough to contain the
-     *                          windowed data's entire life cycle, from window-start through window-end,
-     *                          and for the entire grace period.
-     * @return an instance of a {@link  SessionBytesStoreSupplier}
-     */
-    public static SessionBytesStoreSupplier inMemorySessionStore(final String name, final Duration retentionPeriod) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/Stores.java:365]: Just reordering to "correct" position in alignment to the order of the other method.

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java [hunk 2]
```diff
@@ -465,23 +465,17 @@ public void shouldUseSourceSpecificDeserializers() {
         topology.addSink(
             "sink",
             SINK_TOPIC_1,
-            new Serializer<Object>() {
```
Comment (mjsax) [inline streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java:468]: Java8 cleanup

File: streams/src/main/java/org/apache/kafka/streams/state/internals/GlobalStateStoreProvider.java [hunk 3]
```diff
@@ -41,6 +44,11 @@ public GlobalStateStoreProvider(final Map<String, StateStore> globalStateStores)
         if (!store.isOpen()) {
             throw new InvalidStateStoreException("the state store, " + storeName + ", is not open.");
         }
+        if (store instanceof TimestampedKeyValueStore && queryableStoreType instanceof QueryableStoreTypes.KeyValueStoreType) {
+            return (List<T>) Collections.singletonList(new ReadOnlyKeyValueStoreFacade((TimestampedKeyValueStore<Object, Object>) store));
+        } else if (store instanceof TimestampedWindowStore && queryableStoreType instanceof QueryableStoreTypes.WindowStoreType) {
+            return (List<T>) Collections.singletonList(new ReadOnlyWindowStoreFacade((TimestampedWindowStore<Object, Object>) store));
+        }
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/GlobalStateStoreProvider.java]: Note, change is actually part of #6661 and just added here to make the tests pass -- #6661 should be merged first.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/StreamThreadStateStoreProvider.java [hunk 4]
```diff
@@ -56,7 +58,13 @@ public StreamThreadStateStoreProvider(final StreamThread streamThread) {
...
                             " because the store is not open. The state store may have migrated to another instances.");
                 }
-                stores.add((T) store);
+                if (store instanceof TimestampedKeyValueStore && queryableStoreType instanceof QueryableStoreTypes.KeyValueStoreType) {
+                    stores.add((T) new ReadOnlyKeyValueStoreFacade((TimestampedKeyValueStore<Object, Object>) store));
+                } else if (store instanceof TimestampedWindowStore && queryableStoreType instanceof QueryableStoreTypes.WindowStoreType) {
+                    stores.add((T) new ReadOnlyWindowStoreFacade((TimestampedWindowStore<Object, Object>) store));
+                } else {
+                    stores.add((T) store);
+                }
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/StreamThreadStateStoreProvider.java]: Note, change is actually part of #6661 and just added here to make the tests pass -- #6661 should be merged first.

**Comment by mjsax** [comment]
Java11 failed with env issue. Java8 passed. Not triggering retest as Java11 seems to be broken.

File: streams/src/test/java/org/apache/kafka/streams/integration/StoreUpgradeIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,815 @@
...
+        final Properties props = props();
+        kafkaStreams = new KafkaStreams(streamsBuilderForOldStore.build(), props);
+        kafkaStreams.start();
+
+        final LinkedList<KeyValue<Integer, Object>> expectedResult = new LinkedList<>();
+
+        expectedResult.add(KeyValue.pair(1, 1L));
+        processKeyValueAndVerifyPlainCount(1, expectedResult);
+
+        expectedResult.remove(0);
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/integration/StoreUpgradeIntegrationTest.java]: nit: I'm having a hard time maintaining a clear mental picture of what the expected result looks like as this test progresses. Can you consider using a literal list for each `processKeyValueAndVerifyPlainCount` instead?

**Review by vvcephei — APPROVED** [review]
Thanks, @mjsax !

**Comment by mjsax** [comment]
Updated this

**Review by vvcephei — APPROVED** [review]
Thanks, @mjsax , I know that was a bunch of monotonous work. It's so much clearer to read now!

File: streams/src/test/java/org/apache/kafka/streams/state/internals/StreamThreadStateStoreProviderTest.java [hunk 6]
```diff
@@ -142,6 +160,31 @@ public void shouldFindKeyValueStores() {
...
+    }
+
+    @Test
+    public void shouldFindTimestampedKeyValueStores() {
+        mockThread(true);
+        final List<ReadOnlyKeyValueStore<String, ValueAndTimestamp<String>>> tkvStores =
+            provider.stores("timestamped-kv-store", QueryableStoreTypes.timestampedKeyValueStore());
+        assertEquals(2, tkvStores.size());
+        for (final ReadOnlyKeyValueStore<String, ValueAndTimestamp<String>> store: tkvStores) {
+            assertThat(store, instanceOf(ReadOnlyKeyValueStore.class));
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/StreamThreadStateStoreProviderTest.java]: nit: do we want to add an assertion ` assertThat(store, instanceOf(TimestampedKeyValueStore.class))` here?

File: streams/src/test/java/org/apache/kafka/streams/state/internals/StreamThreadStateStoreProviderTest.java [hunk 7]
```diff
@@ -142,6 +160,31 @@ public void shouldFindKeyValueStores() {
...
+    }
+
+    @Test
+    public void shouldNotFindKeyValueStores() {
+        mockThread(true);
+        final List<ReadOnlyKeyValueStore<String, ValueAndTimestamp<String>>> tkvStores =
+            provider.stores("kv-store", QueryableStoreTypes.timestampedKeyValueStore());
+        assertEquals(2, tkvStores.size());
+        for (final ReadOnlyKeyValueStore<String, ValueAndTimestamp<String>> store: tkvStores) {
+            assertThat(store, instanceOf(ReadOnlyKeyValueStore.class));
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/StreamThreadStateStoreProviderTest.java]: nit: do we want to add an assertion 
 `assertThat(store, not(instanceOf(TimestampedKeyValueStore.class)));` here?

File: streams/src/test/java/org/apache/kafka/streams/state/internals/StreamThreadStateStoreProviderTest.java [hunk 8]
```diff
@@ -142,6 +160,31 @@ public void shouldFindKeyValueStores() {
...
+        final List<ReadOnlyKeyValueStore<String, ValueAndTimestamp<String>>> tkvStores =
+            provider.stores("timestamped-kv-store", QueryableStoreTypes.timestampedKeyValueStore());
+        assertEquals(2, tkvStores.size());
+        for (final ReadOnlyKeyValueStore<String, ValueAndTimestamp<String>> store: tkvStores) {
+            assertThat(store, instanceOf(ReadOnlyKeyValueStore.class));
+        }
+    }
+
+    @Test
+    public void shouldNotFindKeyValueStores() {
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/StreamThreadStateStoreProviderTest.java]: typo in the test name? Looks to me it should be `shouldNotFindTimestampedKeyValueStores`

**Review by bbejeck — COMMENTED** [review]
Thanks @mjsax, just a couple of minor comments, otherwise LGTM

File: streams/src/test/java/org/apache/kafka/streams/state/internals/StreamThreadStateStoreProviderTest.java (hunk 8 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/StreamThreadStateStoreProviderTest.java]: The test name is actually correct, but the test is broken in this version. Rebasing based on #6661 fixed the issue.

**Comment by mjsax** [comment]
Rebase to resolve merge conflicts. Also address latest Github comments. Added new test for in-memory store upgrades. Those fail atm, exposing the bug as fixed in https://github.com/apache/kafka/pull/6667#discussion_r282106131

Jenkins is expected to fail -- need to rebase this after #6667 is merged

**Comment by guozhangwang** [comment]
retest this please

**Comment by mjsax** [comment]
@guozhangwang I need to rebase this first... Test are expected to fail atm.

**Comment by mjsax** [comment]
Rebased this. Call for review.

**Comment by bbejeck** [comment]
Java 8 passed, Java 11 failed with `integration.kafka.api.ConsumerTopicCreationTest.testAutoTopicCreation[brokerTopicCreation=true, consumerTopicCreation=true`

retest this please

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java [hunk 9]
```diff
@@ -735,6 +729,135 @@ public void shouldReturnAllStores() {
...
 
+    @Test
+    public void shouldReturnCorrectStoreTypeOnly() {
+        final String keyValueStoreName = "keyValueStore";
+        final String timestampedKeyValueStoreName = "keyValueTimestampStore";
+        final String windowStoreName = "windowStore";
+        final String timestampedWindowStoreName = "windowTimestampStore";
+        final String sessionStoreName = "sessionStore";
+        final String globalKeyValueStoreName = "globalKeyValueStore";
+        final String globalTimestampedKeyValueStoreName = "globalKeyValueTimestampStore";
```
Comment (bbejeck) [inline streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java:749]: one small nit: should this test include in-memory stores?

**Review by bbejeck — APPROVED** [review]
Thanks @mjsax just one minor comment/question otherwise LGTM

**Comment by guozhangwang** [comment]
retest this please

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java (hunk 9 above)
Comment (mjsax) [reply streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java:749]: Sure. Why not.

**Pull request merged**